docs/config.md
# Configuração
A configuração pode ser feita com o yaml exemplo abaixo:

```yaml
port:     8080,
hostname: localhost
loglevel: info
oidc:
    hostname: localhost
    path:     /auth
credentialCache:
    expiration: 31
    cleanup:    60
credentials:
    type:   csv
    path:   authzman
    reload: 60
jwt:
    rsaPrivateFile: /auth/extauth.rsa,
    localIssuer:    tlsgw.local
    kid:
    tokenLifetime:  60
    issuers:
        name1:
            iss: tlsgw.local
            url: file:///auth/extauth.rsa.pub
```

## port
Porta TCP onde o autenticador vai esctar

## hostname
Hostname a ser utilizado ao abrir a porta

## loglevel
info: informações de boot e tokens assinados/rejeitados

error: Erros de operação

debug: todos os passos da autenticação

## oidc
Parametros para o OIDC atrás do GW
### hostname
endpoint do emissord de tokens
### path
Caminho do endpoint de emissão de tokens

## credentialCache
Atenção! o tempo de expiração deve ser menor que o de limpeza e ambos devem ser menores que o tempo de expiraçãod do token
### expiration
Tempo em segundos para uma jwt assinado expirar no GW

### cleanup
Tempo em segundos para limpeza do cache de tokens

## credentials
Base de credenciais podendo ser CSV ou json

### type
csv ou s3

### reload
intervalo de reconciliação da base de credenciais
## configuração com S3
### bucket
Nome do bucket com as credenciais
### key
caminho no bucket do json de credenciais
### region
Região S3

## Configuracão com CSV
### path
Caminho do csv com as credenciais

## jwt:
Configuração do emissor de tokens
### rsaPrivateFile
Chave privada para emissão

### localIssuer
Hostname a ser utilizado como issuer

### kid
key identitiy, hint para o GW encontrar a chave pública de validação no JWKS
### tokenLifetime
Tempo de expiração do token
### issuers
Configuração dos issuers permitidos de passar pelo GW sem mTLS

Formato de cada entrada segue o padrão
```yaml
Nome:
    iss: issuer configurado no token
    url: caminho do jwks, pode ser file:// ou https://
```

# Credentials
A base de credenciais pode ter dois formatos csv local ou json S3

## CSV
O primeiro campo é o fingerprint do certificado, o segundo é o escopo das audiences e o ultimo as audiences separadas por |

Exemplo:
```csv
"b49d1cdd5a34b98290cd21deb1fc630e101b85f278d9632b60e82ee52263f59a","scope-a","httpbin-get|helloworld"
"b49d1cdd5a34b98290cd21deb1fc630e101b85f278d9632b60e82ee52263f59a","scope-b","httpbin-post|goodbyworld"
"91ac3e2fcbef50072be0d98d8bc242876a850286eb4484103402ed35cec63847","scope-a","httpbin-get|helloworld"
"91ac3e2fcbef50072be0d98d8bc242876a850286eb4484103402ed35cec63847","scope-b","httpbin-post|goodbyworld"
```



## JSON
O Json é uma estrutrua contendo fingerprint|scopes. scopes contem name|audiences, audiences é uma lista

Exemplo:
```json
{
    "fingerprint": "b49d1cdd5a34b98290cd21deb1fc630e101b85f278d9632b60e82ee52263f59a",
    "scopes": [
{
        "name": "scope-a",
        "audiences": [
            "httpbin-get",
            "helloworld"
        ]
    }
,{
        "name": "scope-b",
        "audiences": [
            "httpbin-post",
            "goodbyworld"
        ]
    }
]
}
{
    "fingerprint": "91ac3e2fcbef50072be0d98d8bc242876a850286eb4484103402ed35cec63847",
    "scopes": [
{
        "name": "scope-a",
        "audiences": [
            "httpbin-get",
            "helloworld"
        ]
    }
,{
        "name": "scope-b",
        "audiences": [
            "httpbin-post",
            "goodbyworld"
        ]
    }
]
}
```

# Configuração do Envoy

Para o correto funcionamento do Gw, o envoy deve ser configurado para enviar as solicitacoes de autenticacao para ele

abaixo segue um exemplo de configuracao do envoy:
```yaml
...
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
      - name: envoy.http_connection_manager
        config:
          stat_prefix: egress_http
          rds:
            route_config_name: local_route
            config_source:
              path: /conf/routes.yaml
          forward_client_cert_details: "sanitize_set"
          set_current_client_cert_details:
            subject: True
          access_log:
            name: envoy.file_access_log
            config:
              path: /dev/stdout
          http_filters:
          - name: envoy.ext_authz
            config:
              grpc_service:
                envoy_grpc:
                  cluster_name: extauth
          - name: envoy.router
      tls_context:
        require_client_certificate: false
        common_tls_context:
          validation_context:
            trusted_ca:
              filename: "/ssl/ca.pem"
          tls_certificates:
            - certificate_chain:
                filename: "/ssl/tlsjwt.pem"
              private_key:
                filename: "/ssl/tlsjwt-key.pem"
...

  clusters:
    - name: extauth
      type: STRICT_DNS
      connect_timeout: 0.25s
      http2_protocol_options: {}
      load_assignment:
        cluster_name: extauth
        endpoints:
        - lb_endpoints:
          - endpoint:
              address:
                socket_address:
                  address: extauth
                  port_value: 4000
``` 


cmd/config/config.go
package config

import (
	//"extauth/cmd/jwthandler"
	"fmt"
	"time"

	log "github.com/sirupsen/logrus"
	"github.com/spf13/viper"
)

// DBConf Estrutura para informacoes de tipo e acesso ao provedor das credencias
type DBConf struct {
	// Informacoes necessarias para acesso ao banco que armazena as credenciais
	Options map[string]string
	// Tipo de Banco que esta sendo acessado: csv, s3, dynamo, etc...
	DBType string
}

// PermissionDBConf Estrutura para informacoes de Permissao do provedor das credencias
// INFO O nome PermissionDBConf nao ficou legal, ja que as permissoes estao na estrutura DBConf
type PermissionDBConf struct {
	// Estrutura com informacoes de Type e Options
	Config DBConf
	// Informacoes de quanto tempo as informacoes serao consideradas validas e intervalo de busca
	CacheInterval time.Duration
	CacheClean    time.Duration
}

//type issuerConf interface {
//	iss() string
//	jwks() jwthandler.Jwks
//}

// jwtConf - Estrutura que armazena informacoes de configuracao do JWT
type jwtConf struct {
	RsaPrivateFile string
	LocalIssuer    string
	TokenLifetime  time.Duration
	Issuers        []IssuerConf
	Kid            string
}

// Options - Estrutura com todas as configuracoes necessarias para a execucao
// da aplicacao.
type Options struct {
	Loglevel      string
	Port          int
	Hostname      string
	Oidc          OidcConf
	IgnorePaths   []string
	PermissionDB  *PermissionDBConf
	JwtConf       *jwtConf
	EnableOptions bool
	AuthHeader    string
	ClaimString   string
}

// OidcConf - Estrutura que armazena dados do OIDC
type OidcConf struct {
	Hostname string
	Path     string
}

//// fatal -  kill program
//func fatal(err error) {
//	if err != nil {
//		log.Fatal(err)
//	}
//}
// IssuerConf - Estrutura que armazena dados do Issuer e a URL de acesso a este
type IssuerConf struct {
	Issuer string
	// INFO O padrao na comunicade eh sempre usar a variavel URL toda em caixa
	// alta
	Url string
}

// DefaultConf Retorna uma estrutura em Viper com as configuracoes padroes
func DefaultConf() (v1 *viper.Viper, err error) {
	// Definindo Padroes e lendo arquivo de configuracao
	v1, err = ReadConfig("extauth", map[string]interface{}{
		"port":     8080,
		"hostname": "localhost",
		"loglevel": "debug",

		"oidc": map[string]interface{}{
			"hostname": "localhost",
			"path":     "/auth",
		},
		"IgnorePaths": map[string]interface{}{
			"none": "none",
		},

		"credentialCache": map[string]interface{}{
			"expiration": 31,
			"cleanup":    60,
		},
		"credentials": map[string]string{
			"type":   "csv",
			"path":   "/auth/authzman",
			"reload": "60",
		},
		"jwt": map[string]interface{}{
			"rsaPrivateFile": "/auth/extauth.rsa",
			"localIssuer":    "tlsgw.local",
			"kid":            "",
			"tokenLifetime":  60,
			"authHeader":     "authorization",
			"claimString":    "aud",
			"issuers": map[string]interface{}{
				"name1": map[string]interface{}{
					"iss": "tlsgw.local",
					"url": "file:///auth/extauth.rsa.pub",
				},
			},
		},
	})
	return
}

// ReadConfig - Associa os dados de tipo Mapa para a estrutura Viper e retorna-o
func ReadConfig(filename string, defaults map[string]interface{}) (*viper.Viper, error) {

	// Cria a estrutura Viper
	v := viper.New()

	// Copia os dados do mapa defaults para a estrutura Viper
	for key, value := range defaults {
		v.SetDefault(key, value)
	}

	// Incrementa outra informacoes
	v.SetConfigName(filename)
	v.AddConfigPath(".")
	v.AutomaticEnv()

	// Retorna a estrutura construida
	err := v.ReadInConfig()
	return v, err
}

// BuildOptions - Metodo principal para preenchimento da estrutura options
// Carrega as configuracoes padroes
// Instancia um tipo PermissionDBConf e define as configuracoes de leitura de
// provedor de credenciais
// Instancia um tipo jwtConf e define configuracoes inicias do JWT
func BuildOptions() (Options, error) {
	// Retorna uma estrutura Viper com os dados de configuracoes default
	v1, err := DefaultConf()
	if err != nil {
		panic(fmt.Errorf("Error when reading config: %v\n", err))
	}

	// Cria-se uma estrutura do tipo Options e inicia-se o seu preenchimento de
	// informacoes
	var opt Options
	opt.Hostname = v1.GetString("hostname")
	opt.Port = v1.GetInt("port")

	opt.Loglevel = v1.GetString("loglevel")
	//
	// Config PermissionDB
	//var cc PermissionDBConf

	// Cria uma nova instancia e associa as configuracoes de permissao do
	// provedor
	// A instancia eh instanciada vazia
	// Este sera preenchido ao ler as credenciais
	opt.PermissionDB = new(PermissionDBConf)

	// Preenche o ramo PermissionDB da estrutura opt com o tipo e as informacoes
	// de acesso ao
	// provedor que armazena as credenciais
	switch v1.GetString("credentials.type") {
	case "csv":
		//path := v1.GetString("credentials.path")
		//var param map[string]interface{}
		if path := v1.GetString("credentials.path"); len(path) < 2 {
			log.Fatalf("Erro analisando config do provedor de credenciais %v", err)
		} else {
			param := make(map[string]string)
			param["CsvPath"] = path
			param["interval"] = v1.GetString("credentials.reload")
			opt.PermissionDB.Config = DBConf{param, "csv"}
		}
	case "s3":
		param := make(map[string]string)
		param["bucket"] = v1.GetString("credentials.bucket")
		param["key"] = v1.GetString("credentials.key")
		param["region"] = v1.GetString("credentials.region")
		//interval := v1.GetInt("credentials.reload")
		//if interval < 10 {
		//	log.Fatal("Intervalo de recarga de credenciais não pode ser < 10")
		//}
		param["interval"] = v1.GetString("credentials.reload")
		opt.PermissionDB.Config = DBConf{param, "s3"}
	case "mongo":
		param := make(map[string]string)
		param["uri"] = v1.GetString("credentials.uri")
		param["database"] = v1.GetString("credentials.database")
		opt.PermissionDB.Config = DBConf{param, "mongo"}
		log.Debug("using mongo with %s, %s", v1.GetString("credentials.uri"), v1.GetString("credentials.database"))
	default:
		log.Fatal("Nenhum provedor de credenciais configurado")
	}

	//Configuracoes de AsyncDb

	// Preenche as informacoes de quanto tempo as informacoes obtidas serao
	// consideradas validas
	opt.PermissionDB.CacheClean = time.Duration(v1.GetInt64("credentialCache.cleanup")) * time.Minute
	opt.PermissionDB.CacheInterval = time.Duration(v1.GetInt64("credentialCache.expiration")) * time.Minute

	// Verifica se as configuracoes lidas sao validas para o funcionamento desta
	// aplicacao
	if opt.PermissionDB.CacheInterval > opt.PermissionDB.CacheClean || opt.PermissionDB.CacheInterval < 30 {
		log.Fatal("Tempo de vida do cache minimo 30s, tempo de limpeza deve ser superior ao tempo de vida")
	}

	// JWT Config
	//var jwtConf jwtConf
	// inicia-se a definicao de informacoes de configuracao do JWT que sera
	// emitido
	opt.JwtConf = new(jwtConf)
	// Chave de assinatura dos tokens emitidos pelo GW
	opt.JwtConf.RsaPrivateFile = v1.GetString("jwt.rsaprivatefile")
	opt.JwtConf.LocalIssuer = v1.GetString("jwt.localIssuer")
	opt.JwtConf.Kid = v1.GetString("jwt.kid")
	opt.JwtConf.TokenLifetime = time.Duration(v1.GetInt64("jwt.tokenLifetime"))

	// Adiciona a estrutura opt os caminhos que serao ignorados, descritos pela
	// configuracao lida anteriormente
	ignorePaths := v1.GetStringMap("ignorePaths")
	for e := range ignorePaths {
		// captura um path
		path := v1.GetString(fmt.Sprintf("ignorePaths.%s", e))
		// Adiciona ao final do vetor IgnorePaths
		opt.IgnorePaths = append(opt.IgnorePaths, path)
		log.Debugf("ignore path: %s", path)
	}
	//var issAllow  []IssuerConf

	// Cria um vetor termporario de lista de issuers
	issuers := v1.GetStringMap("jwt.issuers")

	// Cria um vetor do tipo IssuerConf
	var ic []IssuerConf

	// preenche o vetor IssuerConfig com informacoes de Issuer e URL e adiciona
	for k := range issuers {
		iss := v1.GetString(fmt.Sprintf("jwt.issuers.%s.iss", k))
		url := v1.GetString(fmt.Sprintf("jwt.issuers.%s.url", k))
		// INFO Voce esta definindo opt.JwtConf.Issuers aqui e logo que termina
		// o For. Remova a linha depois do for pra evitar processamento
		// desnecessario
		opt.JwtConf.Issuers = append(ic, IssuerConf{iss, url})
	}

	// INFO instrucao desnecessaria
	opt.JwtConf.Issuers = ic

	// Define informacoes de OIDC
	opt.Oidc = OidcConf{v1.GetString("oidc.hostname"), v1.GetString("oidc.path")}

	// Define informacoes de Auth e Claim
	opt.AuthHeader = v1.GetString("jwt.authHeader")
	log.Debugf("authHeader: %s", opt.AuthHeader)
	opt.ClaimString = v1.GetString("jwt.claimString")
	log.Debugf("claimString: %s", opt.ClaimString)
	//opt.IgnoreList

	//fatal(err)

	// Issuer usado pelo GW
	//localIssuer := v1.GetString("jwt.localissuer")
	//log.Debugf("local Issuer: %s", localIssuer)

	// Retorna o arquivo de configuracao construido
	// Possui informacoes de acesso ao provedor de credencial, JWT, etc.
	return opt, err
}
